packages/lit-html/src/disconnectable-directive.ts

```diff
@@ -0,0 +1,384 @@
+ * if the parent list becomes empty as a result, removes the parent from its
+ * parent, and so forth up the tree when that causes subsequent parent lists to
+ * become empty.
+ */
+const removeFromParent = (child: DisconnectableParent) => {
+  let parent = child._$parent;
+  let children = parent?._$disconnetableChildren;
+  children?.delete(child);
+  while (children !== undefined && children.size === 0) {
+    child._$disconnetableChildren = undefined;
+    if (parent !== undefined) {
+      children = parent._$disconnetableChildren!;
+      children.delete(child);
+    } else {
+      break;
+    }
+    child = parent;
+    parent = child._$parent;
+  }
+};
+
+/**
+ * Sets the connected state on any directives contained within the committed
+ * value of this part (i.e. within a TemplateInstance or iterable of NodeParts)
+ * and runs their `disconnectedCallback`/`reconnectedCallback`s, as well as
+ * within any directives stored on the NodePart (when `valueOnly` is false).
+ *
+ * `isClearingValue` should be passed as `true` on a top-level part that is
+ * clearing itself, and not as a result of recursively disconnecting directives
+ * as part of a `clear` operation higher up the tree. This both ensures that any
+ * directive on this NodePart that produced a value that caused the clear
+ * operation is not disconnected, and also serves as a performance optimization
+ * to avoid needless bookkeeping when a subtree is going away; when clearing a
+ * subtree, only the top-most part need to remove itself from the parent.
+ *
+ * `fromPartIndex` is passed only in the case of a partial `_clear` running as a
+ * result of truncating an iterable.
+ *
+ * Note, this method will be patched onto NodePart instances and called from the
+ * core code when parts are cleared or the connection state is changed by the
+ * user.
+ */
+function setNodePartConnected(
+  this: NodePart,
+  isConnected: boolean,
+  isClearingValue = false,
+  fromPartIndex = 0
+) {
+  const value = this._value;
+  const children = this._$disconnetableChildren;
+  if (children !== undefined) {
+    if (isClearingValue) {
+      if (Array.isArray(value)) {
+        // Iterable case: Any NodeParts created by the iterable should be
+        // disconnected and removed from this NodePart's disconnectable
+        // children (starting at `fromPartIndex` in the case of truncation)
+        for (let i = fromPartIndex; i < value.length; i++) {
+          setChildrenConnected(value[i], isConnected);
+          removeFromParent(value[i]);
+        }
+      } else if (value != null) {
+        // TemplateInstance case: If the value has disconnectable children (will
+        // only be in the case that it is a TemplateInstance), we disconnect it
+        // and remove it from this NodePart's disconnectable children
+        setChildrenConnected(value as DisconnectableParent, isConnected);
+        removeFromParent(value as DisconnectableParent);
```

Comment:
should `removeFromParent` be called when `isConnected` is true?

Reply:
Ah, so if `isClearingValue` is true, `isConnected` is always assumed to be false. To make this clearer, I changed it to just pass `false` for `isConnected` when inside the `isClearingValue` case.

It's basically a tri-state of valid values:
* `isClearingValue === true` && `isConnected === false` - value is clearing
* `isClearingValue === false` && `isConnected === false` - part is being manually disconnected
* `isClearingValue === false` && `isConnected === true` - part is being manually reconnected

It sort of points out that these could really just be two separate functions. But I sorta only wanted to patch one method onto `NodePart`... 🤷‍♂️ WDYT?

Reply:
sounds fine as-is